File: src/libqstylefactory.cpp
#include <QList>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QStyle>
#include <QStyleFactory>
#include <qstylefactory.h>
#include "libqstylefactory.h"
#include "libqstylefactory.hxx"

QStyleFactory* QStyleFactory_new(const QStyleFactory* other) {
    return new QStyleFactory(*other);
}

QStyleFactory* QStyleFactory_new2(QStyleFactory* other) {
    return new QStyleFactory(std::move(*other));
}

void QStyleFactory_CopyAssign(QStyleFactory* self, QStyleFactory* other) {
    *self = *other;
}

void QStyleFactory_MoveAssign(QStyleFactory* self, QStyleFactory* other) {
    *self = std::move(*other);
}

libqt_list /* of libqt_string */ QStyleFactory_Keys() {
    QList<QString> _ret = QStyleFactory::keys();
    // Convert QList<> from C++ memory to manually-managed C memory
    libqt_string* _arr = static_cast<libqt_string*>(malloc(sizeof(libqt_string) * (_ret.size() + 1)));
    for (qsizetype i = 0; i < _ret.size(); ++i) {
        QString _lv_ret = _ret[i];
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _lv_b = _lv_ret.toUtf8();
        libqt_string _lv_str;
        _lv_str.len = _lv_b.length();
        _lv_str.data = static_cast<const char*>(malloc(_lv_str.len + 1));
        memcpy((void*)_lv_str.data, _lv_b.data(), _lv_str.len);
        ((char*)_lv_str.data)[_lv_str.len] = '\0';
        _arr[i] = _lv_str;
    }
    libqt_list _out;
    _out.len = _ret.size();
    _out.data = static_cast<void*>(_arr);
    return _out;
}

QStyle* QStyleFactory_Create(const libqt_string param1) {
    QString param1_QString = QString::fromUtf8(param1.data, param1.len);
    return QStyleFactory::create(param1_QString);
}

void QStyleFactory_Delete(QStyleFactory* self) {
    delete self;
}

File: src/libqstylefactory.zig
const C = @import("qt6c");
const std = @import("std");

/// https://doc.qt.io/qt-6/qstylefactory.html
pub const qstylefactory = struct {
    /// New constructs a new QStyleFactory object.
    ///
    /// ``` other: ?*C.QStyleFactory ```
    pub fn New(other: ?*anyopaque) ?*C.QStyleFactory {
        return C.QStyleFactory_new(@ptrCast(other));
    }

    /// New2 constructs a new QStyleFactory object and invalidates the source QStyleFactory object.
    ///
    /// ``` other: ?*C.QStyleFactory ```
    pub fn New2(other: ?*anyopaque) ?*C.QStyleFactory {
        return C.QStyleFactory_new2(@ptrCast(other));
    }

    /// CopyAssign shallow copies `other` into `self`.
    ///
    /// ``` self: ?*QStyleFactory, other: ?*QStyleFactory ```
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        C.QStyleFactory_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    ///
    /// ``` self: ?*QStyleFactory, other: ?*QStyleFactory ```
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        C.QStyleFactory_MoveAssign(@ptrCast(self), @ptrCast(other));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstylefactory.html#keys)
    ///
    /// ``` allocator: std.mem.Allocator ```
    pub fn Keys(allocator: std.mem.Allocator) [][]const u8 {
        const _arr: C.struct_libqt_list = C.QStyleFactory_Keys();
        const _str: [*]C.struct_libqt_string = @ptrCast(@alignCast(_arr.data));
        defer {
            for (0.._arr.len) |_i| {
                C.libqt_string_free(@ptrCast(&_str[_i]));
            }
            C.libqt_free(_arr.data);
        }
        const _ret = allocator.alloc([]const u8, _arr.len) catch @panic("Memory allocation failed");
        for (0.._arr.len) |_i| {
            const _data = _str[_i];
            const _buf = allocator.alloc(u8, _data.len) catch @panic("Memory allocation failed");
            @memcpy(_buf, _data.data[0.._data.len]);
            _ret[_i] = _buf;
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qstylefactory.html#create)
    ///
    /// ``` param1: []const u8 ```
    pub fn Create(param1: []const u8) ?*C.QStyle {
        const param1_str = C.struct_libqt_string{
            .len = param1.len,
            .data = @constCast(param1.ptr),
        };
        return C.QStyleFactory_Create(param1_str);
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: ?*C.QStyleFactory ```
    pub fn QDelete(self: ?*anyopaque) void {
        C.QStyleFactory_Delete(@ptrCast(self));
    }
};
